nlu training data docs updates (#6724)
RasaHQ/rasa: docs/docs/nlu-training-data.mdx

sidebar_label: NLU Training Data
title: NLU Training Data
description: Read more about how to format training data with Rasa NLU for open source natural language processing.
abstract: NLU training data stores structured information about user messages.
---

The goal of NLU (Natural Language Understanding) is to extract structured information from user messages. This usually includes the user's [intent](glossary.mdx#intent) and any 
[entities](glossary.mdx#entities) their message contains. You can
add extra information such as **[regular expressions](#regular-expression-features)** and **[lookup tables](#lookup-tables)** to your
training data to help the model identify intents and entities correctly. 

## Training Examples

NLU training data consists of example user utterances categorized by
intent. 
To make it easier to use your intents, give them names that relate to what the user wants to accomplish with that intent, keep them in lowercase, and avoid spaces and special characters. 

:::note

> The `/` symbol is reserved as a delimiter to separate retrieval intents from response text identifiers. Make sure not

Reviewer: ```suggestion
The `/` symbol is reserved as a delimiter to separate [retrieval intents](glossary.mdx#retrieval-intent) from response text identifiers. Make sure not
```

to use it in the name of your intents.

:::

> ## Entities

Reviewer: Do we not have a quick section about intents?

Reviewer: i'll resolve this for now as its the same structure as the training data format page

When deciding which entities you need to extract, think about what information your assistant needs for its user goals. The user might provide additional pieces of information that you don't need for any user goal; you don't need to extract these as entities.

See the [training data format](./training-data-format.mdx) for details on how to annotate entities in your training data.

## Synonyms

Synonyms map extracted entities to a value other than the literal text extracted. 
You can use synonyms when there are multiple ways users refer to the same
thing. Think of the end goal of extracting an entity, and figure out from there which values should be considered equivalent. 

Let's say you had an entity `account` that you use to look up the user's balance. One of the possible account types is "credit". Your users also refer to their "credit" account as "credit
account" and "credit card account".

> In this case, you could define "credit card account" and "credit account" as
> **synonyms** to "credit":

Reviewer: ```suggestion
In this case, you could define "credit card account" and "credit account" as
synonyms to "credit":
```

```yaml-rasa
nlu:
- synonym: credit
  examples: |
    - credit card account
    - credit account
```

Then, if either of these phrases is extracted as an entity, it will be

> mapped to the **value** `credit`.

Reviewer: ```suggestion
mapped to the value `credit`.
```

> :::note Needs Training Examples

Reviewer: ```suggestion
:::note Provide Training Examples
```

Synonym mapping only happens **after** entities have been extracted.
That means that your training examples should include the synonym examples
(`credit card account` and `credit account`) so that the model will learn to
recognize these as entities and replace them with `credit`.
:::

See the [training data format](./training-data-format.mdx) for details on how to include synonyms in your training data.

## Regular Expressions

You can use regular expressions to improve intent classification and
entity extraction in combination with the [`RegexFeaturizer`](components.mdx#regexfeaturizer) and [`RegexEntityExtractor`](components.mdx#regexentityextractor) components in the pipeline.

### Regular Expressions for Intent Classification

You can use regular expressions to improve intent classification by including the `RegexFeaturizer` component in your pipeline. When using the `RegexFeaturizer`, a regex does not act as a rule for classifying an intent. It only provides a feature that the intent classifier will use
to learn patterns for intent classification.
Currently, all intent classifiers make use of available regex features.

> The name of a regex in this case is a human readable description. It can help you remember what a regex is used for, and it is the title of the corresponding pattern feature. It does not have to match any intent or entity name. A regex for greeting might look like this:

Reviewer 2: I'd say example of regex for such a general intent like greet is a bit dangerous. We actually don't want people to use regexes to identify intents like greet

Reviewer 2: actually regex that matches one word is pointless, because it creates exactly the same feature as word level CountVectorsFeaturizer

```yaml-rasa
nlu:
- regex: help
  examples: |
    - \bhelp\b
```

The intent being matched could be `greet`,`help_me`, `assistance` or anything else.

Try to create your regular expressions in a way that they match as few
words as possible. E.g. using `\bhelp\b` instead of `help.*`, as the
later one might match the whole message whereas the first one only
matches a single word.

> :::note Needs Training Examples
> The `RegexFeaturizer` provides features to the intent classifier, but it doesn't predict the intent directly. Include enough examples containing the regular expression so that the intent classifier can learn ot use the regular expression feature.

Reviewer: ```suggestion
:::note Provide Training Examples
The `RegexFeaturizer` provides features to the intent classifier, but it doesn't predict the intent directly. Include enough examples containing the regular expression so that the intent classifier can learn to use the regular expression feature.
```

:::

### Regular Expressions for Entity Extraction

If your entity has a deterministic structure, you can use regular expressions in one of two ways:

> 1. As features for the [`RegexFeaturizer`](components.mdx#regexfeaturizer) component in the pipeline.

Reviewer 2: "features for featurizer" sounds strange

> 1. As features for the [`RegexFeaturizer`](components.mdx#regexfeaturizer) component in the pipeline.

Reviewer: Lets do these number sections as headers so that everything under it renders correctly 
```suggestion
#### 1. Regular expressions as features for the [`RegexFeaturizer`](components.mdx#regexfeaturizer) component in the pipeline.
```

  :::

  Regex features for entity extraction
  are currently only supported by the `CRFEntityExtractor` and `DIETClassifier` components. Other entity extractors, like
  `MitieEntityExtractor` or `SpacyEntityExtractor`, won't use the generated
  features and their presence will not improve entity recognition for
  these extractors.

2. For rule-based entity extraction using the [`RegexEntityExtractor`](components.mdx#regexentityextractor) component in the pipeline.

  When using the `RegexEntityExtractor`, the name of the regular expression should
  match the name of the entity you want to extract.
  For example, you could extract account numbers of 10-12 digits by including this regular expression and at least two annotated examples in your training data:

  ```yaml-rasa
  nlu:
  - regex: account_number
    examples: |
      - \d{10,12}
  - intent: inform
    examples: |
      - my account number is [1234567891](account_number)
      - This is my account number [1234567891](account_number)
  ```

>   Whenever a user message contains a sequence of 10-12 digits, it will be extracted as an `account_number` entity. `RegexEntityExtractor` doesn't require training examples to learn to extract the entity, but you do need at least two annotated examples of the entity so that the NLU model can register it as an entity at training time.

Reviewer: not rendering correctly

## Lookup Tables

Lookup tables are lists of words used to generate
case-insensitive regular expression patterns. They can be used in the same ways as [regular expressions](#regular-expressions) are used, in combination with the [`RegexFeaturizer`](components.mdx#regexfeaturizer) and [`RegexEntityExtractor`](components.mdx#regexentityextractor) components in the pipeline.

You can use lookup tables to help extract entities which have a known set of possible values. Keep your lookup tables as specific as possible. For example, to extract country names, you could add a lookup table of all countries in the world:

```yaml-rasa
nlu:
- lookup: country
  examples: |
    - Afghanistan
    - Albania
    - ...
    - Zambia
    - Zimbabwe

> ```

Reviewer: can we still provide a path to a file?

Author: no, but the lookup could go in its own file.

Reviewer: coolio, thanks

:::

> Annotating words as custom entities to words allow you to define certain concepts in your training data.

Reviewer: ```suggestion
Annotating words as custom entities allows you to define certain concepts in your training data.
```
